dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckMetaValidationHook.java

```diff
@@ -0,0 +1,203 @@
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.tracker.TrackerIdentifier;
+import org.hisp.dhis.tracker.TrackerImportStrategy;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.tracker.validation.TrackerImportValidationContext;
+import org.springframework.stereotype.Component;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class PreCheckMetaValidationHook
+    extends AbstractTrackerDtoValidationHook
+{
+    @Override
+    public int getOrder()
+    {
+        return 2;
+    }
+
+    @Override
+    public void validateTrackedEntity( ValidationErrorReporter reporter, TrackedEntity tei )
+    {
+        TrackerImportValidationContext context = reporter.getValidationContext();
+
+        OrganisationUnit organisationUnit = context.getOrganisationUnit( tei.getOrgUnit() );
+
+        if ( organisationUnit == null )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1011 )
+                .addArg( reporter ) );
+        }
+
+        TrackedEntityType entityType = context.getTrackedEntityType( tei.getTrackedEntityType() );
+        if ( entityType == null )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1005 )
+                .addArg( tei.getTrackedEntityType() ) );
+        }
+    }
+
+    @Override
+    public void validateEnrollment( ValidationErrorReporter reporter, Enrollment enrollment )
+    {
+        TrackerImportValidationContext context = reporter.getValidationContext();
+        TrackerImportStrategy strategy = context.getStrategy( enrollment );
+
+        OrganisationUnit organisationUnit = context.getOrganisationUnit( enrollment.getOrgUnit() );
+        if ( organisationUnit == null )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1070 )
+                .addArg( enrollment.getOrgUnit() ) );
+        }
+
+        Program program = context.getProgram( enrollment.getProgram() );
+        if ( program == null )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1069 )
+                .addArg( enrollment.getProgram() ) );
+        }
+
+        if ( (program != null && organisationUnit != null) && !program.hasOrganisationUnit( organisationUnit ) )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1041 )
+                .addArg( organisationUnit )
+                .addArg( program )
+                .addArg( program.getOrganisationUnits() ) );
+        }
+
+        //TODO: Change program is not allowed?
```

Comment:
No, this is an error, it would break the model completely

Reply:
Ok :+1: